File: app/src/main/java/com/yeraygarcia/recipes/database/dao/AisleDao.java
package com.yeraygarcia.recipes.database.dao;

import android.arch.lifecycle.LiveData;
import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Query;

import com.yeraygarcia.recipes.database.UUIDTypeConverter;
import com.yeraygarcia.recipes.database.entity.Aisle;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Dao
public abstract class AisleDao extends BaseDao<Aisle> {

    @Query("DELETE FROM aisle")
    public abstract void deleteAll();

    @Query("SELECT * FROM aisle")
    public abstract LiveData<List<Aisle>> findAll();

    @Query("DELETE FROM aisle WHERE id NOT IN (:ids)")
    abstract void deleteIfIdNotIn(List<UUID> ids);

    public void deleteIfNotIn(List<Aisle> entities) {
        List<UUID> ids = new ArrayList<>();
        for (Aisle entity : entities) {
            ids.add(entity.getId());
        }
        deleteIfIdNotIn(ids);
    }
}

File: app/src/main/java/com/yeraygarcia/recipes/database/dao/AisleDao.kt
package com.yeraygarcia.recipes.database.dao

import android.arch.lifecycle.LiveData
import android.arch.persistence.room.Dao
import android.arch.persistence.room.Query
import com.yeraygarcia.recipes.database.entity.Aisle
import com.yeraygarcia.recipes.database.entity.Tag
import java.util.*

@Dao
abstract class AisleDao : BaseDao<Aisle>() {

    @Query("DELETE FROM $TABLE")
    abstract fun deleteAll()

    @Query("DELETE FROM $TABLE WHERE id IN (:ids)")
    abstract fun deleteByIds(ids: List<UUID>)

    @Query("SELECT * FROM $TABLE")
    abstract fun findAll(): LiveData<List<Aisle>>

    @Query("DELETE FROM $TABLE WHERE id NOT IN (:ids)")
    internal abstract fun deleteIfIdNotIn(ids: List<UUID>)

    fun deleteIfNotIn(entities: List<Aisle>) {
        val ids = ArrayList<UUID>()
        for (entity in entities) {
            ids.add(entity.id)
        }
        deleteIfIdNotIn(ids)
    }

    @Query("SELECT * FROM $TABLE")
    abstract fun findAllRaw(): List<Aisle>

    @Query("SELECT * FROM $TABLE WHERE :lastSync = :lastSync")
    abstract fun findModified(lastSync: Long): List<Aisle>

    companion object {
        const val TABLE = "aisle"
    }
}
